### PR TITLE
Automaton refactor

### PR DESCRIPTION
**Pull request by baxtersa**

### CODE REVIEW

**Comment by baxtersa** [comment]
Ahhh, hold on. This is WIP. Global programs aren't compiling.

**Comment by baxtersa** [comment]
This causes switch mods to be inserted into automata leaves, which can cause larger (but semantically equivalent) automata. Switch Mods basically become meta fields when generating a table from a global FDD, so we can simply erase them before generating the table, yielding the same table as before without this change.

File: lib/Frenetic_NetKAT_Compiler.ml [hunk 1]
```diff
@@ -526,17 +541,21 @@ module Pol = struct
...
         | Some ing ->
-            Frenetic_NetKAT_Optimize.(mk_and (Test (Switch s2)) (mk_not ing))
-            |> mk_filter
-      in
-      mk_big_seq [filter_loc s1 p1; Dup; post_link ]
-    | VLink _ -> assert false (* SJS / JNF *)
+          Frenetic_NetKAT_Optimize.(mk_and (Test (Switch s2)) (mk_not ing))
+          |> mk_filter in
+      let link = mk_seq (mk_mod (Switch s2)) (mk_mod (Location (Physical p2))) in
+      mk_big_seq [filter_loc s1 p1; link; Dup; post_link]
```
Comment (smolkaj) [inline lib/Frenetic_NetKAT_Compiler.ml]: This change brakes the global compiler. You're conflating the parts of the program that need to be implemented by the network topology (`link`) with the parts that need to be implemented by the switches (`filter_loc s1 p2`, `post_link`, etc...). But when we extract a local program, we want only the switch stuff, not the topology stuff. You're erasing some of the topology stuff by deleting switch field modifications in `erase_switch_fields`, but you're not erasing the port field modifications.

Comment (smolkaj) [inline lib/Frenetic_NetKAT_Compiler.ml]: If you want a NetKAT automaton that's equivalent to the input program, the right way to go about this is to insert an additional `Dup`:
```
mk_big_seq [filter_loc s1 p1; Dup; link; Dup; post_link]
```
Now topology operations and switch operations are cleanly separated. You end up with a bipartite automaton, where one partition encodes the behavior of the switches and another one the behavior of the topology.
To extract a local program from this, you simply throw away all topology states and only keep the switch states. All switch states can then be merged into a single FDD as before.
By the way, this is described fairly well (I think) in our compiler paper: https://www.cs.cornell.edu/~jnfoster/papers/netkat-compiler.pdf

Comment (baxtersa) [inline lib/Frenetic_NetKAT_Compiler.ml]: Thanks for the feedback. This is what I was intending.

Still getting the hang of this part of the paper/code :).

**Comment by baxtersa** [comment]
@smolkaj , can you double check that I understand this correctly?
With the additional `Dup`, we're now implementing true `Link` semantics during automata construction. 

This yields a bipartite automaton as described by the paper. To extract the local FDD as before, we can filter the topology states (I'm testing whether the `Switch` field is modified at the `Leaf`, is there a better way?), and merge the remaining local states into a single FDD by forwarding all the state pointers to skip over the topology states. This requires `Link` states to have a single outgoing edge, which I believe is the case, right?

**Comment by smolkaj** [comment]
> I'm testing whether the Switch field is modified at the Leaf, is there a better way?

Because the automaton is bipartite, you know that every other state (along any path through the graph) is a topology state. I exploited this to clean up the code a bit; the transformation can be implemented as a map over the automaton states in pre-order. I think the code now reflects very closely what we wrote in the paper.

**Comment by smolkaj** [comment]
I'll go ahead and merge this

**Pull request merged**